apps/prairielearn/src/pages/instructorAssessmentQuestions/components/InstructorAssessmentQuestionsTable.tsx

```diff
@@ -1,127 +1,442 @@
 
-  function maxPointsText({
-    max_auto_points,
-    max_manual_points,
-    points_list,
-    init_points,
+  const handleEditQuestion = ({
+    question,
+    alternativeGroup,
+    zoneNumber,
+    alternativeGroupNumber,
+    alternativeNumber,
   }: {
-    max_auto_points: number | null;
-    max_manual_points: number | null;
-    points_list: number[] | null;
-    init_points: number | null;
-  }) {
-    if (max_auto_points || !max_manual_points) {
-      switch (assessmentType) {
-        case 'Exam':
-          return (points_list || [max_manual_points])
-            .map((p) => (p ?? 0) - (max_manual_points ?? 0))
-            .join(',');
-        case 'Homework':
-          return `${(init_points ?? 0) - (max_manual_points ?? 0)}/${max_auto_points}`;
-        default:
-          assertNever(assessmentType);
+    question: ZoneQuestionJson | QuestionAlternativeJson;
+    alternativeGroup?: ZoneQuestionJson;
+    zoneNumber: number;
+    alternativeGroupNumber: number;
+    alternativeNumber?: number;
+  }) => {
+    setEditQuestionModalState({
+      type: 'edit',
+      question,
+      alternativeGroup,
+    });
+    setSelectedQuestionPosition({
+      zoneNumber,
+      alternativeGroupNumber,
+      alternativeNumber: alternativeNumber ?? undefined,
+    });
+    setQidValidationError(''); // Clear any previous validation errors
+  };
+
+  const handleAddQuestion = (zoneNumber: number) => {
+    setEditQuestionModalState({
+      type: 'create',
+      question: { id: '' } as ZoneQuestionJson,
+    });
+    setSelectedQuestionPosition({
+      zoneNumber,
+      alternativeGroupNumber: mappedQuestions[zoneNumber - 1].questions.length + 1,
+    });
+    setQidValidationError(''); // Clear any previous validation errors
+  };
+
+  const handleUpdateQuestion = async (
+    updatedQuestion: ZoneQuestionJson | QuestionAlternativeJson,
+  ) => {
+    if (!updatedQuestion.id) return;
+    if (!selectedQuestionPosition) return;
+    if (editQuestionModalState.type === 'closed') return;
+
+    // Check if QID already exists in the assessment
+    const isOriginalQid =
+      editQuestionModalState.type === 'edit' &&
+      updatedQuestion.id === editQuestionModalState.question.id;
+    const isDuplicateQid =
+      !isOriginalQid &&
+      mappedQuestions.some((zone) =>
+        zone.questions.some((q) => {
+          if (q.id === updatedQuestion.id) return true;
+          return q.alternatives?.some((alt) => alt.id === updatedQuestion.id) ?? false;
+        }),
+      );
+
+    if (isDuplicateQid) {
+      setQidValidationError('QID already exists in this assessment');
+      return;
+    }
+
+    let questionData: StaffAssessmentQuestionRow | null = null;
+    if (updatedQuestion.id !== editQuestionModalState.question.id) {
+      const params = new URLSearchParams({ qid: updatedQuestion.id });
+      const res = await fetch(`${window.location.pathname}/question.json?${params.toString()}`, {
```

Comment:
All our new clientside code is doing `useQuery`. I don't see a reason to stop doing that.